### PR TITLE
peering: prevent peering in same partition

### PR DESCRIPTION
**Pull request by acpana**
### Description
We want to prevent a `peering` from being created in the same partition (ENT) or cluster (OSS).

We do this by looking for an intersection (any) between the server addresses encoded in the token and those of the current cluster's.

### CODE REVIEW

**Comment by freddygv** [comment]
Hey Alex, server addresses are not specific to a partition, they're specific to a datacenter. Checking whether server addresses match would prevent users from peering two different partitions in the same Consul datacenter.

Another wrinkle is that addresses embedded in a token may not be the same ones that come from `GetServerAddresses()` now that this PR is merged: https://github.com/hashicorp/consul/pull/13844

Could the peer ID help instead?

**Comment by acpana** [comment]
thanks for the guidance @freddygv ! how is this?

**Review by mkeeler — APPROVED** [review]
LGTM.

_Food for thought_

Should we allow peering with any partition within the same cluster. We can already export services across partitions and it seems less than desirable to duplicate those within memdb.

**Review by kisunji — CHANGES_REQUESTED** [review]
I think you accidentally deleted some lines with the rebase

File: agent/rpc/peering/service.go [hunk 1]
```diff
@@ -399,6 +402,21 @@ func (s *Server) Establish(
 	return resp, nil
 }
 
+// validatePeeringInPartition makes sure that we don't create a peering in the same partition. We do so by checking
+// to see if the current server's server addresses slice has any intersection with the slice from the token.
```
Comment (kisunji) [inline agent/rpc/peering/service.go:406]: comment should be updated

File: agent/rpc/peering/service.go [hunk 2]
```diff
@@ -399,6 +402,21 @@ func (s *Server) Establish(
...
+func (s *Server) validatePeeringInPartition(remotePeerID, partition string) error {
+	_, peering, err := s.Backend.Store().PeeringReadByID(nil, remotePeerID)
+	if err != nil {
+		return fmt.Errorf("cannot validate peering in partition: %w", err)
```
Comment (kisunji) [inline agent/rpc/peering/service.go:410]: should just return err instead of wrapping (misleading error message)

Comment (acpana) [reply agent/rpc/peering/service.go:410]: I think it's worth saying the "why" the `PeeringReadByID` is made in the first place.

As a user here, the UX of the "why" (eg `cannot validate peering in partition: error_msg`) let's me know that there is a validation issue with my  request.

Comment (kisunji) [reply agent/rpc/peering/service.go:411]: In that case, a general rule of thumb I like is:
Wrapping error messages should contextualize where the error came from (and avoid self-referencing their containing functions).

In this case, the failure is `PeeringReadByID`, so it's misleading to prematurely wrap the error.
```suggestion
		return fmt.Errorf("cannot read peering by ID: %w", err)
```

The actual context should be wrapped by the caller
```go
	if err := s.validatePeeringInPartition(tok.PeerID, req.Partition); err != nil {
		return nil, fmt.Errorf("cannot validate peering in partition: %w", err)
	}
```

File: agent/rpc/peering/service.go [hunk 3]
```diff
@@ -399,6 +402,22 @@ func (s *Server) Establish(
...
+// the remotePeerID from the PeeringToken and looking up for a peering in the partition. If there is one and the
+// request partition is the same, then we are attempting to peer within the partition, which we shouldn't.
+func (s *Server) validatePeeringInPartition(remotePeerID, partition string) error {
+	_, peering, err := s.Backend.Store().PeeringReadByID(nil, remotePeerID)
+	if err != nil {
+		return fmt.Errorf("cannot validate peering in partition: %w", err)
+	}
+
+	if peering != nil && peering.Partition == partition {
+		return fmt.Errorf("cannot create a peering within the same partition (ENT) or cluster (OSS)")
```
Comment (kisunji) [inline agent/rpc/peering/service.go:415]: Can we better describe what the error is? This applies to the PR body as well. It's unclear what the error condition is

Comment (kisunji) [reply agent/rpc/peering/service.go:415]: nvm I re-read a bunch of times and it makes sense as-is!

**Review by kisunji — APPROVED** [review]
minor comments but LGTM

File: agent/rpc/peering/service.go [hunk 4]
```diff
@@ -354,6 +354,9 @@ func (s *Server) Establish(
 	if err = validatePeer(peeringOrNil, true); err != nil {
 		return nil, err
 	}
+	if err := s.validatePeeringInPartition(tok.PeerID, req.Partition); err != nil {
```
[the remark this answers is not in the document]
Comment (acpana) [reply agent/rpc/peering/service.go:350]: added a comment right above 🥇

**Pull request merged**